README
NAME
    MongoDB::Web - MongoDB + Moose + URI

SYNOPSIS
     package WebPage;
     use Moose;
     extends 'MongoDB::Web::Resource';
     # inherits the 'uri' attribute

     # this attribute is a "Property"
     # it will be stored in mongodb
     has 'title' => (
         traits => [qw( Property )],
         is => 'rw',
         isa => 'Str',
     );

     # this attribute is not a Property
     # it won't be stored
     has 'is_fresh' => (
         is => 'rw',
         isa => 'Bool',
     );

     # this property store an ArrayRef of WebPage URIs
     has outlinks => (
         traits => [qw( Property )],
         is => 'rw',
         isa => 'ArrayRefOfURI',
         uri_isa => 'WebPage',
         coerce => 1,
     );

     package main;
     use MongoDB::Web::Store;

     my $store = MongoDB::Web::Store->new(
         database => MongoDB::Connection->new(host => 'localhost', port => 27017)->test_database
     );

     # new resources
     my $page1 = WebPage->new(
         uri => 'http://example.org/1',
         outlinks => [ 'http://example.org/2' ],
     );
     $store->save($page1);

     my $page2 = WebPage->new(
         uri => 'http://example.org/2',
         outlinks => [ 'http://example.org/1' ],
     );
     $store->save($page2);
 
     # load resources
     my $loaded = $store->load_by_uri('http://example.org/1'); # this is page1 

     my $cursor = $store->load_referers( WebPage => outlinks => $page2 );
     $loaded = $cursor->next; # this is $page1

DESCRIPTION
    MongoDB Object Mapper designed to store Web Resources. The objects are
    identified by their canonical URIs. The scope of this module is actually
    larger than the Web as by default any kind of URIs is accepted. See
    set_alternate_canonical_uri_maker to change this behavior.

METHODS
  $class->canonical_uri( $uri )
    This method is used internaly to validate the URIs and to make them
    canonical. It is critical to store the URIs in their canonical form as
    they are used as primary keys. This method is used for the validation
    and the coercion of the URI and ArrayRefOfURI attributes.

  $class->set_alternate_canonical_uri_maker( sub { my $uri = shift; ...; return $uri; } )
    You can provide your own canonical_uri method as a replacement of the
    standard one.

AUTHOR
    Antoine Imbert, "<antoine.imbert at gmail.com>"

LICENSE AND COPYRIGHT
    Copyright 2010 Antoine Imbert.

    This program is free software; you can redistribute it and/or modify it
    under the terms of either: the GNU General Public License as published
    by the Free Software Foundation; or the Artistic License.

    See http://dev.perl.org/licenses/ for more information.



Changes
Revision history for MongoDB-Web

0.01    Date/time
        First version, released on an unsuspecting world.



t/WebPage.pm
package WebPage;
use Moose;
extends 'MongoDB::Web::Resource';
with 'MongoDB::Web::Store::Shortcuts';

has 'title' => (
    traits => [qw( Property )],
    is => 'rw',
    isa => 'Str',
);

# this attribute has no Property trait
# it won't be stored
has 'is_fresh' => (
    is => 'rw',
    isa => 'Bool',
);

has outlinks => (
    traits => [qw( Property )],
    is => 'rw',
    isa => 'ArrayRefOfURI',
    uri_isa => 'WebPage',
    coerce => 1,
);

1;


lib/MongoDB/Web/Store.pm
package MongoDB::Web::Store;
use Moose;
use MongoDB::OID;

use MongoDB::Web;
use MongoDB::Web::Resource;
use MongoDB::Web::Cursor;

has 'database' => (
    is => 'rw',
    isa => 'MongoDB::Database',
    required => 1,
);

=head1 NAME

=head1 SYNOPSIS

 my $store = MongoDB::Web::Store->new(
     database => MongoDB::Connection->new(host => 'localhost', port => 27017)->test_web
 );
 
=head1 DESCRIPTION

=head1 METHODS

=head2 $class->new( database => $mdb )

Takes a MongoDB::Database in argument.

=head2 $self->load_by_uri( $class => $uri )

=cut

sub load_by_uri {
    my $self = shift;
    my $class = shift or die 'class required';
    my $uri = shift or die 'uri required';

    $uri = MongoDB::Web->canonical_uri($uri);
    my $c = $self->class_to_collection($class);

    my $doc = $c->find_one(
        { uri => $uri },
    );
    return unless $doc;
    return $class->new_from_document($doc);
}

=head2 $self->load_by_mongodb_id( $class => $id )

=cut

sub load_by_mongodb_id {
    my $self = shift;
    my $class = shift or die 'class required';
    my $id = shift or die 'id required';

    my $oid = MongoDB::OID->new( value => $id );
    my $c = $self->class_to_collection($class);

    my $doc = $c->find_one(
        { _id => $oid },
    );

    return unless $doc;
    return $class->new_from_document($doc);
}

=head2 $self->load( $resource => $property );

If the property type is URI then return the resource.
If the property type is ArrayRefOfURI then return a cursor.

=cut

sub load {
    my $self = shift;
    my $resource = shift or die 'resource required';
    die "must be a MongoDB::Web::Resource"
        unless $resource->isa('MongoDB::Web::Resource');
    my $property = shift or die 'property required';

    my $attr = $resource->meta->find_attribute_by_name( $property )
        or die 'property not found';
    my $type = $attr->type_constraint->name;
    
    die 'type must be URI or ArrayRefOfURI'
        unless $type eq 'URI' || $type eq 'ArrayRefOfURI';

    my $class = $attr->uri_isa;
    die "class must inherit from MongoDB::Web::Resource"
        unless $class->isa('MongoDB::Web::Resource');

    if ($type eq 'URI') {
        my $uri = $attr->get_value($resource);
        return $self->load_by_uri( $class => $uri );
    }
    else {
        my $uris = $attr->get_value($resource);
        return $self->find($class => { uri => { '$in' => $uris } });
    }
}

=head2 $self->load_referers( $class => $property => $resource )

Return a cursor.

=cut

sub load_referers {
    my $self = shift;
    my $class = shift or die 'class required';
    die "class $class must inherit from MongoDB::Web::Resource"
        unless $class->isa('MongoDB::Web::Resource');
    my $property = shift or die 'property required';
    my $resource = shift or die 'resource required';
    die "must be a MongoDB::Web::Resource"
        unless $resource->isa('MongoDB::Web::Resource');

    my $uri = $resource->uri;
    return $self->find( $class => { $property => $uri } );
}

=head2 $self->find( $class => $query )

Return a cursor.

=cut

sub find {
    my $self = shift;
    my $class = shift or die 'class required';
    my $query = shift or die 'query required';
    my $c = $self->class_to_collection($class);
    my $cursor = $c->find($query);
    return MongoDB::Web::Cursor->new(cursor => $cursor, class => $class);
}

=head2 $self->save( $resource )

=cut

# TODO support options safe ?
sub save {
    my $self = shift;
    my ($resource) = @_;
    die "must be a MongoDB::Web::Resource"
        unless $resource->isa('MongoDB::Web::Resource');
    my $class = ref $resource;
    my $c = $self->class_to_collection($class);
    return $c->update(
        { uri => $resource->uri },
        $resource->document,
        { upsert => 1 }
    );
}

=head2 $self->remove($resource);

=cut

# TODO support option safe ?
sub remove {
    my $self = shift;
    my ($resource) = @_;
    die "must be a MongoDB::Web::Resource"
        unless $resource->isa('MongoDB::Web::Resource');
    my $class = ref $resource;
    my $c = $self->class_to_collection($class);
    return $c->remove({ uri => $resource->uri });
}

=head2 $self->ensure_index( $class => ... )

=cut

sub ensure_index {
    my $self = shift;
    my $class = shift or die 'class required';
    my $c = $self->class_to_collection($class);
    return $c->ensure_index(@_);
}

=head2 $self->drop_index( $class => ... )

=cut

sub drop_index {
    my $self = shift;
    my $class = shift or die 'class required';
    my $c = $self->class_to_collection($class);
    return $c->drop_index(@_);
}

=head1 ACCESSING UNDERLYING MONGODB OBJECTS.

=head2 $self->database

Returns the MongoDB::Database object.

=head2 $self->class_to_collection_name( $class_name )

=cut

# TODO support custom mapping
sub class_to_collection_name {
    my $self = shift;
    my ($class) = @_;
    die "class required" unless $class;
    $class =~ s/::/_/g;
    return lc $class;
}

=head2 $self->class_to_collection( $class_name )

=cut

my %uri_index_ensured;

sub class_to_collection {
    my $self = shift;
    my ($class) = @_;
    my $c_name = $self->class_to_collection_name($class);
    my $c = $self->database->$c_name();

    # don't run this all the times;
    unless ($uri_index_ensured{$c_name}) {
        $c->ensure_index({ uri => 1 }, { unique => 1 });
        $uri_index_ensured{$c_name}++;
    }

    return $c;
}

=head1 AUTHOR

Antoine Imbert, C<< <antoine.imbert at gmail.com> >>

=head1 LICENSE AND COPYRIGHT

Copyright 2010 Antoine Imbert.

This program is free software; you can redistribute it and/or modify it
under the terms of either: the GNU General Public License as published
by the Free Software Foundation; or the Artistic License.

See http://dev.perl.org/licenses/ for more information.

=cut

1;


lib/MongoDB/Web/Cursor.pm
package MongoDB::Web::Cursor;
use Moose;

has 'cursor' => (
    is => 'rw',
    isa => 'MongoDB::Cursor',
    required => 1,
); 

has 'class' => (
    is => 'rw',
    isa => 'Str',
    required => 1,
); 

=head1 METHODS

=head2 $class->new( cursor => ..., class => ... )

=head2 $self->next

=cut

sub next {
    my $self = shift;    
    my $doc = $self->cursor->next;
    return unless $doc;
    my $class = $self->class;
    return $class->new_from_document($doc);
}

=head2 $self->all

=cut

sub all {
    my $self = shift;
    my $class = $self->class;
    return map {
        $class->new_from_document($_)
    } $self->cursor->all;
}

sub _proxy {
    my $self = shift;
    my $mtd = shift;
    $self->cursor->$mtd(@_);
    return $self;
}

=head2 $self->count

=cut

sub count { shift->cursor->count(@_) }

=head2 $self->has_next

=cut

sub has_next { shift->cursor->has_next(@_) }

=head2 $self->reset

=cut

sub reset { shift->cursor->reset(@_) }

=head2 $self->explain

=cut

sub explain { shift->cursor->explain(@_) }

=head2 $self->snapshot

=cut

sub snapshot { shift->cursor->snapshot(@_) }

=head2 $self->skip( ... )

Proxy to MongoDB::Cursor::skip. Returns a MongoDB::Web::Cursor.

=cut

sub skip { shift->_proxy(skip => @_) }

=head2 $self->limit( ... )

Proxy to MongoDB::Cursor::limit. Returns a MongoDB::Web::Cursor.

=cut

sub limit { shift->_proxy(limit => @_) }

=head2 $self->sort( ... )

Proxy to MongoDB::Cursor::sort. Returns a MongoDB::Web::Cursor.

=cut

sub sort { shift->_proxy('sort' => @_) }

=head2 $self->hint( ... )

Proxy to MongoDB::Cursor::hint. Returns a MongoDB::Web::Cursor.

=cut

sub hint { shift->cursor->hint(@_) }

=head1 ACCESSING UNDERLYING MONGODB OBJECTS

=head2 $self->cursor

Returns the underlying MongoDB::Cursor

=cut

1;


lib/MongoDB/Web/Resource.pm
package MongoDB::Web::Resource;
use Moose;
use MongoDB::OID;
use URI;

use MongoDB::Web;
use MongoDB::Web::Property;

use Moose::Util::TypeConstraints;

subtype 'URI',
    as 'Str',
    where { $_ eq MongoDB::Web->canonical_uri($_)},
    message { "$_ is not a canonical URI, set coerce => 1" };

coerce 'URI',
    from 'Str', via { MongoDB::Web->canonical_uri($_) };

subtype 'ArrayRefOfURI',
    as 'ArrayRef[URI]';

coerce 'ArrayRefOfURI',
    from 'ArrayRef[Str]',
    via { [ map { MongoDB::Web->canonical_uri($_) } @$_ ] },
;

has 'uri' => (
    traits => [qw( Property )],
    is => 'ro',
    isa => 'URI',
    required => 1,
    coerce => 1,
);

has '_id' => (
    traits => [qw( Property )],
    is => 'ro',
    isa => 'Object',
);

=head1 NAME

MongoDB::Web::Resource

=head1 SYNOPSIS

=head1 DESCRIPTION

=head1 METHODS

=head2 $class->new( uri => $uri )

=head2 $self->uri

=head1 ACCESSING UNDERLYING MONGODB OBJECTS

=head2 $self->mongodb_id

Returns the MongoDB ID of the document.

=cut

sub mongodb_id {
    my $self = shift;
    my $oid = $self->{_id};
    return $oid ? $oid->value : undef;
}

=head2 $self->document

=cut

sub document {
    my $self = shift;
    my $meta = $self->meta;

    my @properties = 
        grep { $_->does('MongoDB::Web::Property') }
        $meta->get_all_attributes;

    my %document;
    for (@properties) {
        my $name = $_->name;
        my $value = $_->get_value($self);
        next unless defined $value;
        $document{$name} = $value;
    }
    return \%document;
}

=head2 $class->new_from_document( $document )

=cut

sub new_from_document {
    my $class = shift;
    my ($doc) = @_;
    return $class->new($doc);
}

1;


lib/MongoDB/Web/Store/Shortcuts.pm
package MongoDB::Web::Store::Shortcuts;
use Moose::Role;

=head1 SYNOPSIS

 package WebPage;
 use Moose;
 extends 'MongoDB::Web::Resource';
 with 'MongoDB::Web::Store::Shortcuts';

 package main;
 use MongoDB::Web::Store;
 use WebPage;

 my $store = ...;
 WebPage->set_store( $store );

 my $page = WebPage->new( 'http://example.org' );
 $page->save;

=head1 DESCRIPTION

Using this role an setting the store gives you access to these shortcuts methods.

=cut

has store_access => (
    is => 'rw',
    isa => 'Bool',
    default => 1,
);

my %class2store;

=head2 $class->set_store( $store )

Tell this role which store to use for this class.
To prevent error, the store can be set only once. 

=cut

sub set_store {
    my $class = shift;
    my ($store) = @_;
    die 'store required' unless $store;
    die 'store must be a MongoDB::Web::Store'
        unless $store->isa('MongoDB::Web::Store');
    die 'class must be a MongoDB::Web::Resource'
        unless $class->isa('MongoDB::Web::Resource');

    die 'store already set' if $class2store{$class};

    $class2store{$class} = $store;
}

sub _get_store {
    my $class = shift;
    $class = ref $class if ref $class;
    my $store = $class2store{$class}
        or die 'no store: use set_store';
    return $store;
}

=head2 $self->store_access( Bool )

Disable the load, load_referers, save, and remove methods.
This is usefull when you want this object to be render by a view,
to make sure this view does no IO (at least store IO)

=head2 $class->load_by_uri( $uri )

Short for $store->load_by_uri( $class => $uri )

=cut

sub load_by_uri {
    my $class = shift;
    return $class->_get_store->load_by_uri( $class => @_ );
}

=head2 $class->load_by_mongodb_id( $id )

Short for $store->load_by_mongodb_id( $class => $id )

=cut

sub load_by_mongodb_id {
    my $class = shift;
    return $class->_get_store->load_by_mongodb_id( $class => @_ );
}

=head2 $class->find( $query )

Short for $store->find( $class => $query )

=cut

sub find {
    my $class = shift;
    return $class->_get_store->find( $class => @_ );
}

=head2 $class->ensure_index( ... )

Short for $store->ensure_index($class => ...)

=cut

sub ensure_index {
    my $class = shift;
    return $class->_get_store->ensure_index($class => @_);
}

=head2 $class->drop_index( ... )

Short for $store->drop_index($class => ...)

=cut

sub drop_index {
    my $class = shift;
    return $class->_get_store->drop_index($class => @_);
}

=head2 $self->load( $property )

Short for $store->load( $self => $property )

=cut

sub load {
    my $self = shift;
    die 'no store access' unless $self->store_access;
    return $self->_get_store->load( $self => @_ );
}

=head2 $self->load_referers( $class => $property )

Short for $store->load_referers( $class => $property => $self )

=cut

sub load_referers {
    my $self = shift;
    my ($class, $property) = @_;
    die 'no store access' unless $self->store_access;
    return $self->_get_store->load_referers( $class => $property => $self );
}

=head2 $self->save

Short for $store->save($self)

=cut

sub save {
    my $self = shift;
    die 'no store access' unless $self->store_access;
    return $self->_get_store->save($self);
}

=head2 $self->remove

Short for $store->remove($self)

=cut

sub remove {
    my $self = shift;
    die 'no store access' unless $self->store_access;
    return $self->_get_store->remove($self);
}

1;


lib/MongoDB/Web.pm
package MongoDB::Web;
use warnings;
use strict;

use URI;

our $VERSION = '0.01';

=head1 NAME

MongoDB::Web - MongoDB + Moose + URI

=head1 SYNOPSIS

 package WebPage;
 use Moose;
 extends 'MongoDB::Web::Resource';
 # inherits the 'uri' attribute

 # this attribute is a "Property"
 # it will be stored in mongodb
 has 'title' => (
     traits => [qw( Property )],
     is => 'rw',
     isa => 'Str',
 );

 # this attribute is not a Property
 # it won't be stored
 has 'is_fresh' => (
     is => 'rw',
     isa => 'Bool',
 );

 # this property store an ArrayRef of WebPage URIs
 has outlinks => (
     traits => [qw( Property )],
     is => 'rw',
     isa => 'ArrayRefOfURI',
     uri_isa => 'WebPage',
     coerce => 1,
 );

 package main;
 use MongoDB::Web::Store;

 my $store = MongoDB::Web::Store->new(
     database => MongoDB::Connection->new(host => 'localhost', port => 27017)->test_database
 );

 # new resources
 my $page1 = WebPage->new(
     uri => 'http://example.org/1',
     outlinks => [ 'http://example.org/2' ],
 );
 $store->save($page1);

 my $page2 = WebPage->new(
     uri => 'http://example.org/2',
     outlinks => [ 'http://example.org/1' ],
 );
 $store->save($page2);
 
 # load resources
 my $loaded = $store->load_by_uri('http://example.org/1'); # this is page1 

 my $cursor = $store->load_referers( WebPage => outlinks => $page2 );
 $loaded = $cursor->next; # this is $page1

=head1 DESCRIPTION

MongoDB Object Mapper designed to store Web Resources.
The objects are identified by their canonical URIs.
The scope of this module is actually larger than the Web as by default
any kind of URIs is accepted.
See set_alternate_canonical_uri_maker to change this behavior.

=head1 METHODS

=cut

my $canonical_uri_maker = sub {
    my ($uri) = @_;
    return URI->new($uri)->canonical->as_string;
};

=head2 $class->canonical_uri( $uri )

This method is used internaly to validate the URIs and to make them canonical.
It is critical to store the URIs in their canonical form as they are used as primary keys.
This method is used for the validation and the coercion of the URI and ArrayRefOfURI attributes.

=cut

sub canonical_uri {
    my $class = shift;
    my $uri = shift or die 'uri required';
    return $canonical_uri_maker->($uri);
}

=head2 $class->set_alternate_canonical_uri_maker( sub { my $uri = shift; ...; return $uri; } )

You can provide your own canonical_uri method as a replacement of the standard one.

=cut

sub set_alternate_canonical_uri_maker {
    my $class = shift;
    my $maker = shift or die 'maker required';
    die 'make must be a coderef' unless ref $maker eq 'CODE';
    $canonical_uri_maker = $maker;
}

=head1 AUTHOR

Antoine Imbert, C<< <antoine.imbert at gmail.com> >>

=head1 LICENSE AND COPYRIGHT

Copyright 2010 Antoine Imbert.

This program is free software; you can redistribute it and/or modify it
under the terms of either: the GNU General Public License as published
by the Free Software Foundation; or the Artistic License.

See http://dev.perl.org/licenses/ for more information.

=cut

1;
